i6_models/assemblies/conformer/conformer_v2.py

```diff
@@ -0,0 +1,132 @@
+    swap_mhsa_conv: bool = True
+
+
+class ConformerBlockV2(nn.Module):
+    """
+    Conformer block module
+    """
+
+    def __init__(self, cfg: ConformerBlockV2Config):
+        """
+        :param cfg: conformer block configuration with subunits for the different conformer parts
+        """
+        super().__init__()
+        self.ff1 = ConformerPositionwiseFeedForwardV1(cfg=cfg.ff_cfg)
+        self.mhsa = ConformerMHSAV1(cfg=cfg.mhsa_cfg)
+        self.conv = ConformerConvolutionV1(model_cfg=cfg.conv_cfg)
+        self.ff2 = ConformerPositionwiseFeedForwardV1(cfg=cfg.ff_cfg)
+        self.final_layer_norm = torch.nn.LayerNorm(cfg.ff_cfg.input_dim)
+        self.swap_mhsa_conv = cfg.swap_mhsa_conv
+
+    def forward(self, x: torch.Tensor, /, sequence_mask: torch.Tensor) -> torch.Tensor:
+        """
+        :param x: input tensor of shape [B, T, F]
+        :param sequence_mask: mask tensor where 1 defines positions within the sequence and 0 outside, shape: [B, T]
+        :return: torch.Tensor of shape [B, T, F]
+        """
+        x = 0.5 * self.ff1(x) + x  #  [B, T, F]
+        if not self.swap_mhsa_conv:
+            x = self.mhsa(x, sequence_mask) + x  #  [B, T, F]
+            x = self.conv(x) + x  #  [B, T, F]
+        else:
+            x = self.conv(x) + x  # [B, T, F]
+            x = self.mhsa(x, sequence_mask) + x  # [B, T, F]
+        x = 0.5 * self.ff2(x) + x  #  [B, T, F]
+        x = self.final_layer_norm(x)  #  [B, T, F]
+        return x
+
+
+@dataclass
+class ConformerEncoderV2Config(ModelConfiguration):
+    """
+    Attributes:
+        num_layers: Number of conformer layers in the conformer encoder
+        loss_scales: Dictionary defining the loss layers and corresponding loss scales e.g. {6: 0.3, 12:1}
+        frontend: A pair of ConformerFrontend and corresponding config
+        block_cfg: Configuration for ConformerBlockV2
+    """
+
+    num_layers: int
+    loss_scales: Dict[int, float]
+
+    # nested configurations
+    frontend: ModuleFactoryV1
+    block_cfg: ConformerBlockV2Config
+
+
+class ConformerEncoderV2(nn.Module):
+    """
+    Modification compared to ConformerEncoderV1:
+    1. enable adding auxiliary losses with different loss scales
+        (C.f. https://arxiv.org/abs/2102.03216 for more details of auxiliary loss)
+    2. enable swapping the MHSA and Conv module
+        (C.f. https://arxiv.org/abs/2011.10798 for intuition behind swapping)
+    """
+
+    def __init__(self, cfg: ConformerEncoderV2Config):
+        """
+        :param cfg: conformer encoder configuration with subunits for frontend and conformer blocks
+        """
+        super().__init__()
+
+        self.frontend = cfg.frontend()
+        self.module_list = torch.nn.ModuleList([ConformerBlockV2(cfg.block_cfg) for _ in range(cfg.num_layers)])
+        self.output_layers = list(cfg.loss_scales.keys())
+        assert cfg.num_layers in self.output_layers, "The final layer must be included in loss layers"
```

Comment:
when you put this into the forward instead of the class init, you could even say the user is allowed to not access the final layer, and then you do a `min(len(...), max(return_layers))` to determine how long the module loop should work.

Reply:
I don't fully get the point here. If the final layer is not included and user is allowed to not access the final layer, then it would waste the memory to load the layers between [max(loss_layers),  num_layers].

Reply:
Yes, it would waste the memory but not the computation. I am just saying this is an unnecessary limitation in my view. I would just let this be user choice and not enforce it.

Reply:
could you explain more about `min(len(...), max(return_layers))`, I don't understand the idea here

Reply:
What I meant:
```python
        outputs = []
        for i in range(min(len(self.module_list), max(return_layers)):
            x = self.module_list[i](x, sequence_mask)  # [B, T, F']
            if i + 1 in return_layers:
                outputs.append(x)
                ```
But probably much better idea:
```python
        outputs = []
        assert max(return_layers) <= len(self.module_list)
        for i in range(max(return_layers)):
            x = self.module_list[i](x, sequence_mask)  # [B, T, F']
            if i + 1 in return_layers:
                outputs.append(x)
                ```